drivers/dai/intel/alh/alh.c

```diff
@@ -0,0 +1,223 @@
+	alh->params.rate = ALH_RATE_DEFAULT;
+	/* the LSB 8bits are for stream id */
+	alh->params.stream_id = alh_cfg->mapping[0].alh_id & 0xff;
+
+	return 0;
+}
+
+static int dai_alh_trigger(const struct device *dev, enum dai_dir dir,
+			   enum dai_trigger_cmd cmd)
+{
+	LOG_INF("%s cmd %d", __func__, cmd);
+
+	return 0;
+}
+
+static void alh_claim_ownership(void)
+{
+#if !CONFIG_INTEL_ADSP_CAVS
+	uint32_t ALHASCTL = DT_INST_PROP_BY_IDX(0, reg, 0);
+	uint32_t ALHCSCTL = DT_INST_PROP_BY_IDX(0, reg, 1);
+
+	sys_write32(sys_read32(ALHASCTL) | ALHASCTL_OSEL(0x3), ALHASCTL);
+	sys_write32(sys_read32(ALHCSCTL) | ALHASCTL_OSEL(0x3), ALHCSCTL);
+#endif
+}
+
+static void alh_release_ownership(void)
+{
+#if !CONFIG_INTEL_ADSP_CAVS
+	uint32_t ALHASCTL = DT_INST_PROP_BY_IDX(0, reg, 0);
+	uint32_t ALHCSCTL = DT_INST_PROP_BY_IDX(0, reg, 1);
+
+	sys_write32(sys_read32(ALHASCTL) | ALHASCTL_OSEL(0), ALHASCTL);
+	sys_write32(sys_read32(ALHCSCTL) | ALHASCTL_OSEL(0), ALHCSCTL);
+#endif
+}
+
+
+static const struct dai_config *dai_alh_config_get(const struct device *dev, enum dai_dir dir)
+{
+	struct dai_config *params = (struct dai_config *)dev->config;
+	struct dai_intel_alh *dp = (struct dai_intel_alh *)dev->data;
+	struct dai_intel_alh_pdata *alh = dai_get_drvdata(dp);
+
+	params->rate = alh->params.rate;
+	params->channels = alh->params.channels;
+	params->word_size = ALH_WORD_SIZE_DEFAULT;
+
+	return params;
+}
+
+static int dai_alh_config_set(const struct device *dev, const struct dai_config *cfg,
+				  const void *bespoke_cfg)
+{
+	struct dai_intel_alh *dp = (struct dai_intel_alh *)dev->data;
+
+	LOG_INF("%s", __func__);
+
+	alh_claim_ownership();
+
+	if (cfg->type == DAI_INTEL_ALH) {
+		return dai_alh_set_config_tplg(dp, cfg, bespoke_cfg);
+	} else {
+		return dai_alh_set_config_blob(dp, bespoke_cfg);
+	}
+}
+
+static const struct dai_properties *dai_alh_get_properties(const struct device *dev,
+							   enum dai_dir dir, int stream_id)
+{
+	struct dai_intel_alh *dp = (struct dai_intel_alh *)dev->data;
+	struct dai_intel_alh_pdata *alh = dai_get_drvdata(dp);
+	struct dai_properties *prop = &alh->props;
+	uint32_t offset = dir == DAI_DIR_PLAYBACK ?
+		ALH_TXDA_OFFSET : ALH_RXDA_OFFSET;
+
+	prop->fifo_address = dai_base(dp) + offset + ALH_STREAM_OFFSET * stream_id;
+	prop->dma_hs_id = alh_handshake_map[stream_id];
+	prop->stream_id = alh->params.stream_id;
+
+	LOG_INF("%s dai_index %u", __func__, dp->index);
+	LOG_INF("%s fifo %u", __func__, prop->fifo_address);
+	LOG_INF("%s handshake %u", __func__, prop->dma_hs_id);
+
+	return prop;
+}
+
+static int dai_alh_probe(struct dai_intel_alh *dp)
+{
+	struct dai_intel_alh_pdata *alh = dai_get_drvdata(dp);
+
+	LOG_INF("%s", __func__);
+
+	memset(alh, 0, sizeof(*alh));
+
+	return 0;
+}
+
+static int dai_alh_remove(struct dai_intel_alh *dp)
+{
+	LOG_INF("%s", __func__);
+
+	alh_release_ownership();
+
+	return 0;
+}
+
+static int dai_alh_probe_wrapper(const struct device *dev)
```

Comment:
Seems like this "wraps" the probe method just to take and release a lock and increment a refcount?  Seems like you could put that into probe() directly, no?  Nothing else calls dai_alh_probe().

Reply:
Sure, removed all wrapper functions